website/blog/2024/04-23-scrapy-vs-crawlee/index.md

```diff
@@ -0,0 +1,323 @@
+}
+```
+
+Crawlee's storage can be divided into two categories: Request Storage (Request Queue and Request List) and Results Storage (Datasets and Key Value Stores). Both are stored locally by default in the `./storage` directory.
+
+Also, remember that Crawlee, by default, clears its storages before starting a crawler run. This action is taken to prevent old data from interfering with new crawling sessions.
+
+Let's see how Crawlee stores the result:
+
+- You can use local storage with dataset
+
+```js
+import { PlaywrightCrawler } from 'crawlee';
+
+const crawler = new PlaywrightCrawler({
+    requestHandler: async({ page }) => {
+
+        const title = await page.title();
+        const price = await page.textContent('.price');
+    
+        await crawler.pushData({
+            url: request.url,
+            title,
+            price
+        });
+    }
+})
+
+await crawler.run(['http://example.com']);
+```
+
+- Using Key-Value Store
+
+```js
+import { KeyValueStore } from 'crawlee';
+//... Code to crawl the data
+KeyValueStore.setValue('key', { foo: 'bar' });
+```
+
+### Anti-blocking and Fingerprints
+
+In Scrapy, handling anti-blocking strategies like [IP rotation](https://pypi.org/project/scrapy-rotated-proxy/), [user-agent rotation](https://python.plainenglish.io/rotating-user-agent-with-scrapy-78ca141969fe), custom solutions via middleware, and plugins are needed.
+
+Crawlee provides HTTP crawling and [browser fingerprints](https://crawlee.dev/docs/guides/avoid-blocking) with zero configuration necessary; fingerprints are enabled by default and available in `PlaywrightCrawler` and `PuppeteerCrawler` but work with `CheerioCrawler` and the other HTTP Crawlers too. 
+
+### Error handling
+
+Both libraries support error-handling practices like automatic retries, logging, and custom error handling.
+
+In Scrapy, you can handle errors using middleware and [signals](https://docs.scrapy.org/en/latest/topics/signals.html). There are also [exceptions](https://docs.scrapy.org/en/latest/topics/exceptions.html) like `IgnoreRequest`, which can be raised by Scheduler or any downloader middleware to indicate that the request should be ignored. Similarly, a spider callback can raise' CloseSpider' to close the spider.
+
+Scrapy has built-in support for retrying failed requests. You can configure the retry policy (e.g., the number of retries, retrying on particular HTTP codes) via settings such as `RETRY_TIMES`, as shown in the example:
+
+```py
+# In settings.py
+RETRY_ENABLED = True
+RETRY_TIMES = 2  # Number of retry attempts
+RETRY_HTTP_CODES = [500, 502, 503, 504, 522, 524]  # HTTP error codes to retry
+```
+
+In Crawlee, you can also setup your custom error handler. For retries, `maxRetries` controls how often Crawlee will retry a request before marking it as failed. To setup it in code you just need to add the following line of code in your crawler.
+
+```js
+maxRetries = 3 // Crawler will retry three times.
+```
```

Comment:
This is confusing, because it's not valid JS syntax. I'd do at least:

```suggestion
```js
const crawler = new CheerioCrawler({
    maxRetries: 3 // Crawler will retry three times.
    // ...
})
```
```

Reply:
also the option is called `maxRequestRetries`, not just `maxRetries`

Reply:
good catch, mixed it with scrapy. thanks!